### ISSUE TITLE
Support high bit depth NDI

### ISSUE THREAD

**Issue by llamafilm** [CESNET/UltraGrid#120]
I would like to try encoding HDR video from NDI with NVENC.  Proper HDR requires a minimum of 10-bits.  NDI can provide 16-bit 4:2:2 using `NDIlib_FourCC_type_P216`.  Would it be possible to add support for this format?

NVENC doesn't support this input format so we would have to convert to yuv444p16le or p010le.

**Comment by alatteri** [comment]
have you tried `--param lavc-use-codec=yuv444p16le`?

**Comment by llamafilm** [comment]
I could be wrong, but as I understand it, that parameter controls what is being sent to the encoder.  Up-converting an 8-bit source to 16-bit does not interest me.  The NDI input is limited to one of these options:
```
NDIlib_FourCC_type_UYVY
NDIlib_FourCC_type_RGBA
NDIlib_FourCC_type_BGRA
```

**Comment by llamafilm** [comment]
I'm not a C programmer by any means, but this looks pretty easy in the NDI SDK.  You just have to initialize `NDIlib_recv_create_v3()` with  `color_format = NDIlib_recv_color_format_best`.
I was going to try hacking this myself, but on the UltraGrid side, I don't see any 16-bit 4:2:2 options in `codec_t` so I am stuck.

**Comment by MartinPulec** [comment]
Thanks @llamafilm for investigating that, I think that it may be implemented relatively soon.

**Comment by MartinPulec** [comment]
I've added support to current UltraGrid code for the following:

1. basic support for Y216 codec (packed version of P216) - testcard + conversions from/to UYVY
2. capturing NDI to Y216 if NDI passes P216
3. libavcodec conversion from Y216 to AV_PIX_FMT_YUV444P16

I didn't manage to do extensive testing (especially for 2. and 3.) since it is harder to setup so please be patient - all together may not yet play together well now. If you'd be able to test, it would be superb.

**Comment by MartinPulec** [comment]
Hi @llamafilm,

assuming you didn't have time to test because I've just figure out, there was small mistakes that prevented the NDI capture from working. Unfortunately I still didn't have an opportunity to test (I am not able to generate high bit-depth signal now, perhaps it would be possible to add 16-bit NDI display /?/).

**Comment by MartinPulec** [comment]
Displaying 16-bit with NDI is now possible, but interestingly if sending the 16-bit YCbCr, the receive always provides 8-bit even when "best" codec is selected... Not sure if there is some NDI setting how to control that? However from UG perspective it may be working (although I still was not able to test that 16-bit capture).

**Comment by llamafilm** [comment]
Yeah sorry I have not had time to focus on this project lately.

The SDK docs are a little vague, saying P216 works on "most platforms" but I don't know what that means.  I did read anecdotally one person say it only works on Windows but I have not tested.

**Comment by llamafilm** [comment]
I finally had some time to test this.  It seems like the NDI receiver always defaults to Y216 even if the source is UYVY.  Here I am using [Nobe Display](https://timeinpixels.com/nobe-display/) as a source which is supposed to be 16-bit.  This is running on Windows.

```
uv -t ndi:url=192.168.4.3:5963 -d gl
UltraGrid 1.6+ (remotes/origin/HEAD rev a5d85855 built May 11 2021 16:08:49)

Display device   : gl
Capture device   : ndi
Audio capture    : none
Audio playback   : none
MTU              : 9000 B
Video compression: none
Audio codec      : PCM
Network protocol : UltraGrid RTP
Audio FEC        : none
Video FEC        : none

[NAT] Private outbound IPv4 address detected and bound to a non-dynamic port. Consider adding '-N' option for NAT traversal.
Created new RTP session with SSRC 0x363515d5.
GL setup: fullscreen: OFF, deinterlace: OFF
Display initialized-gl
192.168.4.3:5963Video capture initialized-ndi
Created new RTP session with SSRC 0x261b2bf6.
[NDI cap.] Looking for source(s)...
[GL] OpenGL 2.0 is supported...
[NDI cap.] No sources.
[NDI cap.] Looking for source(s)...
Setting GL window size 512x512 (512x512).
[NDI cap.] No sources.
[NDI cap.] Looking for source(s)...
[NDI cap.] Receiving from source: DESIGNARE (Nobe Display NDI connect), URL: 192.168.4.3:5963
No data received.
[NDI cap.] Received video changed: 1920x1080 @25.00p, codec Y216
Assertion failed: video_frame.line_stride_in_bytes == vc_get_linesize(video_frame.xres, out_desc.color_spec), file src/video_capture/ndi.cpp, line 501

UltraGrid has crashed.
```

If I use Ultragrid as the source (`uv.exe -t testcard:1920:1080:24:UYVY -d ndi`), the receiver crashes in exactly the same way, again identifying the source as Y216 even though it should be UYVY this time.

**Comment by llamafilm** [comment]
I had NDI SDK 4.5 installed.  I upgraded to 4.6 and rebuild Ultragrid from latest git but it still crashes the same way.

**Comment by MartinPulec** [comment]
Thanks, I'll take a look at it. The selection of codec, however, is on the NDI library in the end, UltraGrid defaults to `NDIlib_recv_color_format_best`, which is described as:

    This format will try to provide the video in the format that is the closest to native for the incoming
    codec yielding the highest quality. Specifically, this allows for receiving on 16bpp color from many
    sources. 

I don't know what influences the selection, I have rather an opposite problem - even using ` -t testcard:1920:1080:24:Y216` yields UYVY capture for me. Maybe the Windows sender works differently? I'll try to find out. Anyways, you may force the codec with **color=** option.

Anyways, I'd fix the assertion, but there is the problem that I don't know how to push the NDI to provide me Y216 to the receiver to see what stride actually NDI assumes (maybe there is a padding, I may at least not assume that there isn't any).

To sum up, I am unfortunately uncertain how the selection of codec works - isn't there any system setting or so? But from my perspective the use of `NDIlib_recv_color_format_best` as described above seems correct. But it is open for discussion. If you had some more information about how the workflow operates it will be appreciated.

**Comment by MartinPulec** [comment]
I've pushed a change (0a1f2953) that should fix the assert fail - I suppose that the stride is 0 for P216, anyways, because it is a semi-planar format and thus it doesn't make much sense as a single number. I'd be glad if you could check - I had no chance to test the internal conversion (the NDI's semi-planar P216 is converted to packed Y216).

Anyways, if it is true that the input is always encoded as 16-bit as indicated in the already mentioned [article](https://support.newtek.com/hc/en-us/articles/218109667-NDI-Encoding-Decoding), it would indeed be worth considering if not to let 8-bit as default - otherwise it would perhaps always capture 16 bits in 16-bit enabled setups (that I somehow lack). 

Perhaps we can also wait for **NDI 5** which has already been pre-announced to see how it behaves there?

**Comment by llamafilm** [comment]
I think it makes sense to use "best" by default.  What exactly "best" means may depend on the platform.  I heard a rumor that 16-bit "best" only works on Windows, although this is not mentioned in the documentation.  I will probably leave it on best all the time, but perhaps some users might want a flag way to change that.  Maybe someone can test this on slower hardware to see if it matters.

Testing now from testcard source is working fine, both for UYVY and Y216.  But I'm having a problem with Nobe Display.  Ultragrid on Windows detects the source as Y216 and there are strange white artifacts near black and white pixels.  The same problem appears with both GL and Decklink outputs.  The NDI Studio Monitor app looks fine from the same Nobe Display source.  If I switch Nobe Display output to 8-bit then the picture looks fine.

On macOS, Ultragrid detects this same source as UYVY and the image looks fine.  So perhaps the Mac library simply does not support 16-bit (?)  I suspect maybe the Windows Studio Monitor app also doesn't support 16-bit, so that's why it looks fine.  I don't know.

Here is an image showing the artifacts on the golf club and near the top and bottom edge.  There is a watermark because I have not paid for Nobe Display.
<img width="1326" alt="Screen Shot 2021-06-11 at 11 02 25" src="https://user-images.githubusercontent.com/3991046/121730504-9f3db200-caa4-11eb-88aa-81ca6739281a.png">

**Comment by MartinPulec** [comment]
Thanks for tracking this down. I've also tried the NobeOmniScope (trial version) source using SMPTE Color Bar and B/W Gradient and haven't yet managed to reproduce.

But I have another idea - what de/compression do you use? Or can you try different or uncompressed? The thing is that I've recently fixed a bug with NVENC HEVC where the decompressed samples were out-of-range (limited) YCbCr for which the UG conversions were not prepared so it caused overflows. If you don't have current code (14th May) it may be this issue. Of course it may be elsewhere but the manifestation look like really indicating some over/underflow.

**Comment by llamafilm** [comment]
The Nobe Display app should be using the NDI native codec, is that what you mean?  I'm not doing any compression (yet), just outputting straight to GL or Decklink with Ultragrid.

**Comment by MartinPulec** [comment]
Okay, thanks for the info. I didn't use Nobe Display but rather Nobe OmniScope, I guess that Nobe Display requires Davinci Resolve running as a signal source, right? In the OmniScope there is only option for NDI 8 or 16 bits (roughly, the rest is related rather to input signal).

**Comment by llamafilm** [comment]
Ah yes, it relies on Resolve as a source by running as an OFX plugin. The only NDI related option is 8 or 16 bit. In 8-bit mode it works fine but in 16-bit mode I get these artifacts/noise.

**Comment by llamafilm** [comment]
I think I see the problem now.  These artifacts appear on super whites / sub-blacks (outside legal range video).  At least this is how it appears when looking at the same source in Scopebox waveform.  I will see if this can be fixed on the sender side.
NDI Studio Monitor shows these pixels as crushed black/white which is probably the right way to handle it.

**Comment by MartinPulec** [comment]
Yes, this sounds legit. Please let me know if you find something out. I have meanwhile checked that there shouldn't be problem with the out-of-range values in GL display:

    uv -t testcard:1920:1080:25:Y216:pattern=raw=0xffffff7fffffff7f -d gl

The above command is displayed correctly as white (the _raw_ pattern has just been added). The only further place where I think it should be in UltraGrid is in _NDI capture_ the conversion P216->Y216 but this is purely byte shift operations so I wouldn't expect that problem here.

**Comment by MartinPulec** [comment]
Closing for inactivity now. You can let us know if the problem persists and you have some more details to provide.

**Comment by sejkii** [comment]
Hi did you find a solution? I really like to have HDR (10bit) for NDI.
As soon HDR is enabled in windows I get black screen with mouse curser only on obs (NDI obs plugin)on my second pc

**Comment by MartinPulec** [comment]
@sejkii yes, as far as I know this issue is resolved, UltraGrid supports more than 8 bit depth formats. The issue was kept open because the author reported some issues that I was not able to reproduce and didn't get any update.
